### PR TITLE
Rewrite Lock/Delete key range ops to prevent deadlocks

### PR DESCRIPTION
**Pull request by EinKrebs**

### CODE REVIEW

File: qdb/memqdb.go [hunk 1]
```diff
@@ -167,86 +172,51 @@ func (q *MemQDB) UpdateKeyRange(ctx context.Context, keyRange *KeyRange) error {
...
 
 	// Wait until key range will be unlocked
+	success := true
+	lockedKR := ""
 	var locks []*sync.RWMutex
-	for _, l := range q.Locks {
-		l.Lock()
+	for krId, l := range q.Locks {
+		if !l.TryLock() {
+			success = false
```
Comment (Denchick) [inline qdb/memqdb.go:204]: you don't need `success` and `lockedKR`, just do `return spqrerror.Newf(spqrerror.SPQR_KEYRANGE_ERROR, "key range \"%s\" is locked", lockedKR)`

File: qdb/memqdb.go [hunk 2]
```diff
@@ -167,86 +172,51 @@ func (q *MemQDB) UpdateKeyRange(ctx context.Context, keyRange *KeyRange) error {
...
-		defer q.muDeletedKrs.Unlock()
-		delete(q.deletedKrs, id)
-	}()
-
-	q.mu.RLock()
+	q.mu.Lock()
+	defer q.mu.Unlock()
 
-	// Wait until key range will be unlocked
 	if lock, ok := q.Locks[id]; ok {
```
Comment (Denchick) [inline qdb/memqdb.go:181]:
```
lock, ok := q.Locks[id];
if !ok {
    return nil // or error
}
if !lock.TryLock() {
    return spqrerror.Newf(spqrerror.SPQR_KEYRANGE_ERROR, "key range \"%s\" is locked", id)
}
defer lock.Unlock()

```

**Review by Denchick — CHANGES_REQUESTED** [review]

File: Makefile [hunk 3]
```diff
@@ -87,7 +87,8 @@ pooler_run:
 ####################### TESTS #######################
 
 unittest:
-	go test -race ./cmd/... ./pkg/... ./router/... ./qdb/... ./coordinator/... ./yacc/console...
+	go test ./cmd/... ./pkg/... ./router/... ./coordinator/... ./yacc/console...
```
[the remark this answers is not in the document]
Comment (EinKrebs) [reply Makefile:90]: MemQDB test is the only one where we specifically test for races & deadlocks, it doesn't currently make sense with other tests

[the remark this answers is not in the document]
Comment (EinKrebs) [reply Makefile:90]: Also it slows tests down to repeat them 20 times

[the remark this answers is not in the document]
Comment (reshke) [reply Makefile:90]: why do we remove -race option?

[the remark this answers is not in the document]
Comment (Denchick) [reply Makefile:90]: > Also it slows tests down to repeat them 20 times

I thought it was a feature, not a bug. They are still fast enough

**Review by Denchick — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
[the remark this answers is not in the document]
Comment (EinKrebs) [reply Makefile:90]: Tests other than MemQDB are essentially deterministic, there's no point of re-running them

**Review by Denchick — APPROVED** [review]

**Pull request merged**